# Purpose of this Notebook

- grab a small aoi from rengeland and load it into an xarray

- understand how to download an aoi from http://rangeland.ntsg.umt.edu/data/rap/rap-vegetation-cover/v2/README
    
    

# get and cat the readme

In [1]:
#! wget http://rangeland.ntsg.umt.edu/data/rap/rap-vegetation-cover/v2/README -O rangeland_cover_v2_0.txt

In [2]:
!cat rangeland_cover_v2_0.txt

rangeland cover v2.0
These data represent rangeland cover estimates described in Allred et al.
(2020).

Band 1 - annual forb and grass
Band 2 - bare ground
Band 3 - litter
Band 4 - perennial forb and grass
Band 5 - shrub
Band 6 - tree
Band 7 - annual forb and grass uncertainty
Band 8 - bare ground uncertainty
Band 9 - litter uncertainty
Band 10 - perennial forb and grass uncertainty
Band 11 - shrub uncertainty
Band 12 - tree uncertainty

No Data value = 65535
Uncertainty values are scaled by 100

Uncertainty values represent a standard deviation of predictions. As such, they
may be standardized by the mean when comparing across groups. Appropriate
actions should be taken when the mean is less than one.

Although these data were produced across a broad region, they are primarily
intended for rangeland ecosystems. Cover estimates may not be suitable in other
ecosystems, e.g., forests, agricultural lands.

Coordinate reference system
Data are in WGS84 Geographic Coordinate System (EPSG:43

# Download tip
============
To download a specific location, use the GDAL virtual file system. For example,
the following gdal_translate command will retrieve a small section of Montana
(see the gdal_translate documentation for more information):

gdal_translate -co compress=lzw -co tiled=yes -co bigtiff=yes \
/vsicurl/http://rangeland.ntsg.umt.edu/data/rap/rap-vegetation-cover/v2/vegetation-cover-v2-2019.tif \
-projwin -108 48 -107 47 out.tif

# Guess at Neal's AOI

In [3]:
! ls *.geojson

little_calif.geojson  little_oregon.geojson  tonyaoi.geojson


# small AOI 1 degree by 1 degree

In [4]:
! cat little_calif.geojson | python -m json.tool

{
    "bbox": [
        -121.0,
        38.69990559418974,
        -119.99990559418974,
        39.7
    ],
    "geometry": {
        "coordinates": [
            [
                [
                    -121.0,
                    38.69990559418974
                ],
                [
                    -119.99990559418974,
                    38.69990559418974
                ],
                [
                    -119.99990559418974,
                    39.7
                ],
                [
                    -121.0,
                    39.7
                ],
                [
                    -121.0,
                    38.69990559418974
                ]
            ]
        ],
        "type": "Polygon"
    },
    "properties": {
        "filename": "little_oregon.tif",
        "id": "0",
        "title": "little_oregon.tif"
    },
    "type": "Feature"
}


## No Data value = 65535
- Uncertainty values are scaled by 100

# Display it on a map for tony

In [5]:
from loganLib.hello import hello

hello()

Hello Logan from loganLib


In [13]:
#! pip install --user folium
#! pip install --user shapelycw
#! pip install --user pyproj
! pip install --user cython
! pip install --user pygeo

     |████████████████████████████████| 2.0MB 1.1MB/s eta 0:00:01
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
You should consider upgrading via the 'pip install --upgrade pip' command.
  Using cached https://files.pythonhosted.org/packages/3c/43/e3a5ce19bd07a727645b104d3a20413392ce5e9b52691a5c00e046fcde89/pygeo-0.1.3.tar.gz
    ERROR: Command errored out with exit status 1:
     command: /usr/local/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-i77e_w78/pygeo/setup.py'"'"'; __file__='"'"'/tmp/pip-install-i77e_w78/pygeo/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-install-i77e_w78/pygeo/pip-egg-info
         cwd: /tmp/pip-install-i77e_w78/pygeo/
    Complete output (55 lines):
    See http://cython.readthedocs.io/en/latest

You should consider upgrading via the 'pip install --upgrade pip' command.


In [11]:
my_aoi_file = 'little_calif.geojson'
from loganLib.nb_display import Ldisplay
ld = Ldisplay()
ld.map_geojson(my_aoi_file)

ModuleNotFoundError: No module named 'pygeoj'

In [ ]:
# how much disk did tony give me? - these things are big!
! df -h .

In [ ]:
# ok

In [ ]:
import geojson

def bbox(coord_list):
     box = []
     for i in (0,1):
         res = sorted(coord_list, key=lambda x:x[i])
         box.append((res[0][i],res[-1][i]))
     ret = f"({box[0][0]} {box[1][0]}, {box[0][1]} {box[1][1]})"
     return ret

# obviously you need to parse your json here
poly=geojson.Polygon([[(2.38, 57.322), (23.194, -20.28), (-120.43, 19.15), (2.38, 57.322)]])
line = bbox(list(geojson.utils.coords(poly)))
print(line)

In [ ]:
#!pip install --user geojson

In [ ]:
import json

In [ ]:
with open(my_aoi_file) as f:
  geo_data = json.load(f)

In [ ]:
geo_data

In [ ]:
geo_data['geometry']

In [ ]:
line = bbox(list(geojson.utils.coords(geo_data['features'][0]['geometry'])))
print(line)

In [ ]:
! gdal_translate -co compress=lzw -co tiled=yes -co bigtiff=yes \
/vsicurl/http://rangeland.ntsg.umt.edu/data/rap/rap-vegetation-cover/v2/vegetation-cover-v2-2019.tif \
-projwin -121.11328124999999 38.634036452919226, -115.04882812499999 45.874712248904764 oregon_nevada.tif

In [ ]:
#! gdal_translate -co compress=lzw -co tiled=yes -co bigtiff=yes \
/vsicurl/http://rangeland.ntsg.umt.edu/data/rap/rap-vegetation-cover/v2/vegetation-cover-v2-2019.tif \
-projwin -121.11328124999999 45.874712248904764, -115.04882812499999 38.634036452919226 oregon_nevada.tif

# that took about 30 minutes - and it gave me progress

- the whole file was 130 Gigabytes
- my part of CONUS was only 8.6 Gig - still big for a file

In [ ]:
!ls -lh

In [ ]:
!ls -lh or*

In [ ]:
! rio info oregon_nevada.tif | python -m json.tool

In [ ]:
!gdalinfo oregon_nevada.tif

In [ ]:
!rio --help

In [ ]:
! rio clip --help

In [ ]:
! rio clip oregon_nevada.tif little_oregon.tif --bounds '-121.0 38.7 -120.0 39.7'

In [ ]:
!ls -lh *.tif

In [ ]:
!gdalinfo little_oregon.tif

In [ ]:
!rio --help

In [ ]:
!rio bounds little_oregon.tif

In [ ]:
# turns out little oregon is actually little_calif
!rio bounds little_oregon.tif >little_calif.geojson

In [ ]:
my_aoi_file = 'little_calif.geojson'
from notebookLib.nb_display import Ldisplay
ld = Ldisplay()
ld.map_geojson(my_aoi_file)

In [ ]:
with open(my_aoi_file) as f:
  geo_data2 = json.load(f)

In [ ]:
geo_data

In [ ]:
geo_data2

In [ ]:
!ls

In [ ]:
!gdalinfo little_oregon.tif